tests/functional_tests/wallet_scanner.cpp

```diff
@@ -0,0 +1,667 @@
+    const rct::key &legacy_base_spend_pubkey,
+    const crypto::secret_key &legacy_spend_privkey,
+    const crypto::secret_key &legacy_view_privkey,
+    sp::mocks::ClientConnectionPool &conn_pool)
+{
+    std::unordered_map<rct::key, cryptonote::subaddress_index> legacy_subaddress_map{};
+    add_default_subaddresses(legacy_base_spend_pubkey, legacy_view_privkey, legacy_subaddress_map);
+
+    // Default config pointing to updated daemon
+    const sp::scanning::ScanMachineConfig updated_scan_config{
+            .reorg_avoidance_increment  = 1,
+            .max_chunk_size_hint        = 20, // the lower this is, the quicker feedback gets to the user on scanner progress
+            .max_partialscan_attempts   = 0
+        };
+
+    const std::uint64_t updated_pending_chunk_queue_size = std::min(
+        (std::uint64_t)(boost::thread::hardware_concurrency() + 2),
+        static_cast<std::uint64_t>(10));
+
+    const sp::scanning::mocks::AsyncScanContextLegacyConfig config{
+            .pending_chunk_queue_size = updated_pending_chunk_queue_size,
+            .max_get_blocks_attempts  = 3,
+            .trusted_daemon           = true
+        };
+
+    const std::function<bool(const cryptonote::COMMAND_RPC_GET_BLOCKS_FAST::request&, cryptonote::COMMAND_RPC_GET_BLOCKS_FAST::response&)> rpc_get_blocks =
+        [&conn_pool](const cryptonote::COMMAND_RPC_GET_BLOCKS_FAST::request &req, cryptonote::COMMAND_RPC_GET_BLOCKS_FAST::response &res)
+            {
+                return conn_pool.rpc_command<cryptonote::COMMAND_RPC_GET_BLOCKS_FAST>(
+                    sp::mocks::ClientConnectionPool::http_mode::BIN,
+                    "/getblocks.bin",
+                    req,
+                    res);
+            };
+
+    sp::EnoteFindingContextLegacySimple enote_finding_context{legacy_base_spend_pubkey,
+        legacy_subaddress_map,
+        legacy_view_privkey};
+
+    sp::scanning::mocks::AsyncScanContextLegacy scan_context_ledger{config,
+        enote_finding_context,
+        async::get_default_threadpool(),
+        rpc_get_blocks};
+
+    sp::SpEnoteStore user_enote_store{
+            /*refresh_index*/                   start_height == 0 ? 1 : start_height,
+            /*first_sp_allowed_block_in_chain*/ std::uint64_t(-1),
+            /*default_spendable_age*/           CRYPTONOTE_DEFAULT_TX_SPENDABLE_AGE
+        };
+
+    sp::mocks::ChunkConsumerMockLegacy chunk_consumer{legacy_base_spend_pubkey,
+        legacy_spend_privkey,
+        legacy_view_privkey,
+        user_enote_store};
+
+    sp::scanning::ScanContextNonLedgerDummy scan_context_nonledger{};
+
+    bool r = sp::refresh_enote_store(updated_scan_config,
+        scan_context_nonledger,
+        scan_context_ledger,
+        chunk_consumer);
+    CHECK_AND_ASSERT_THROW_MES(r, "Failed to refresh enote store");
+
+    // Now that we're done scanning, we can close all open connections and keep 1 open for more RPC calls
+    conn_pool.close_connections(1);
```

Comment:
Ok basically did the above and thought more on it.. technically all the tests also assume the daemon and wallets aren't shared / in use by other owners too, but they could be in theory. Separately, I also wanted to restructure the tests into a class so I don't need to pass the daemon / wallet / conection pool through nested calls. So I implemented a wrapper accessor to each of them that checks for ownership of the respective mutex before allowing access

FWIW it's not really perfect as is, since the underlying daemon / wallet / connection pool can still be accessed without going through the accessors. Maybe you have more thoughts on a better approach still